### PR TITLE
Label plugin

### PR DESCRIPTION
**Pull request by xwz123**
1. support the command to add or remove kind|priority|sig/* labels
2. clear labels by config when pr source branch changed

### CODE REVIEW

File: prow/gitee-plugins/label/clear.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+	sdk "gitee.com/openeuler/go-gitee/gitee"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"k8s.io/test-infra/prow/gitee"
+)
+
+func (l *label) handleClearLabel(e *sdk.PullRequestEvent, log *logrus.Entry) error {
+	org, repo := gitee.GetOwnerAndRepoByEvent(e)
```
Comment (xwz123) [inline prow/gitee-plugins/label/clear.go:15]: need rebase and modify

File: prow/gitee-plugins/label/label.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+	if ne.IsIssue() {
+		return l.handleIssueCommentEvent(gitee.NewIssueNoteEvent(e), labelMatches, removeLabelMatches, log)
+	}
+
+	return nil
+}
+
+func (l *label) handlePRCommentEvent(e gitee.PRNoteEvent, addMatches, rmMatches [][]string, log *logrus.Entry) error {
+	org, repo := gitee.GetOwnerAndRepoByEvent(e.NoteEvent)
```
Comment (xwz123) [inline prow/gitee-plugins/label/label.go:140]: ditto

File: prow/gitee-plugins/label/label.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+		removeMatchLabels(rmMatches, prLabels, action, log)
+	}
+	if len(addMatches) > 0 {
+		return addMatchLabels(addMatches, prLabels, repoLabels, action, log)
+	}
+	return nil
+}
+
+func (l *label) handleIssueCommentEvent(e gitee.IssueNoteEvent, addMatches, rmMatches [][]string, log *logrus.Entry) error {
+	org, repo := gitee.GetOwnerAndRepoByEvent(e.NoteEvent)
```
Comment (xwz123) [inline prow/gitee-plugins/label/label.go:157]: ditto

File: prow/gitee-plugins/label/note_handler.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+func (pnh *prNoteHandle) getLabels() (map[string]string, error) {
+	labels, err := pnh.client.GetPRLabels(pnh.org, pnh.repo, pnh.number)
+	if err != nil {
+		return nil, err
+	}
+	return labelsTransformMap(labels), nil
+}
+
+func (pnh *prNoteHandle) handleComment() error {
```
Comment (zengchen1024) [inline prow/gitee-plugins/label/note_handler.go:107]: The implementation of handleComment is same as the one of issue

Comment (zengchen1024) [reply prow/gitee-plugins/label/note_handler.go:107]: move this method to noteHandle

File: prow/gitee-plugins/label/note_handler.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	sdk "gitee.com/openeuler/go-gitee/gitee"
+)
+
+type handler interface {
```
Comment (zengchen1024) [inline prow/gitee-plugins/label/note_handler.go:12]: no use, delete

File: prow/gitee-plugins/label/clear.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	if err = l.removeLabels(org, repo, number, needClear, log); err != nil {
+		return err
+	}
+
+	comment := fmt.Sprintf(
+		"This pull request source branch has changed,label(s): %s has been removed.", strings.Join(needClear, ","))
+	return l.ghc.CreatePRComment(org, repo, number, comment)
+}
+
+func (l *label) removeLabels(org, repo string, number int, rms []string, log *logrus.Entry) error {
```
Comment (zengchen1024) [inline prow/gitee-plugins/label/clear.go:39]: can two `removeLabels` be merge as one

File: prow/gitee-plugins/label/label.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+		formattedLabels = append(formattedLabels, fmt.Sprintf(`"%s/*"`, label))
+	}
+	if len(formattedLabels) > 0 {
+		return fmt.Sprintf(
+			"The label plugin will work on %s and %s labels.",
+			strings.Join(formattedLabels[:len(formattedLabels)-1], ", "), formattedLabels[len(formattedLabels)-1],
+		)
+	}
+	return ""
+}
```
Comment (zengchen1024) [inline prow/gitee-plugins/label/label.go:174]: tyep handler interface {
	addLabel(org, repo string, label []string) error
	addComment(comment string) error
	removeLabel(label string) error
	getLabels() (map[string]string, error)
}
type noteHandle struct {
	client       giteeClient
	log          *logrus.Entry
	addLabels    []string
	removeLabels []string
	org          string
	repo         string
	
	h handler
}

func (nh *noteHandle) getRepoLabels() (map[string]string, error) {
	labels, err := nh.client.GetRepoLabels(nh.org, nh.repo)
	if err != nil {
		return nil, err
	}
	return labelsTransformMap(labels), nil
}


func (inh *noteHandle) handleComment() error {
	repoLabels, err := inh.getRepoLabels()
	if err != nil {
		return err
	}
	issueLabels, err := inh.h.getLabels()
	if err != nil {
		return err
	}

	if len(inh.removeLabels) > 0 {
		removeLabels(inh.h.removeLabel, issueLabels, inh.removeLabels, inh.log)
	}

	if len(inh.addLabels) > 0 {
		return addLabels(inh.h, issueLabels, repoLabels, inh.addLabels)
	}
	return nil
}

type issueNoteHandle struct {
	number string
}

func (inh *issueNoteHandle) addLabel(org, repo string, label []string) error {
	return inh.client.AddMultiIssueLabel(org, repo, inh.number, label)
}

func (inh *issueNoteHandle) addComment(org, repo, comment string) error {
	return inh.client.CreateIssueComment(inh.org, inh.repo, inh.number, comment)
}

func (inh *issueNoteHandle) removeLabel(label string) error {
	return inh.client.RemoveIssueLabel(inh.org, inh.repo, inh.number, label)
}

func (inh *issueNoteHandle) getLabels() (map[string]string, error) {
	labels, err := inh.client.GetIssueLabels(inh.org, inh.repo, inh.number)
	if err != nil {
		return nil, err
	}
	return labelsTransformMap(labels), nil
}



	var neh handler
	if ne.IsPullRequest() {
		number := gitee.NewPRNoteEvent(e).GetPRNumber()
		neh = &prNoteHandle{noteHandle: nh, number: number}
	} else if ne.IsIssue() {
		number := gitee.NewIssueNoteEvent(e).GetIssueNumber()
		neh = &issueNoteHandle{noteHandle: nh, number: number}
	} else {
		return nil
	}


	org, repo := ne.GetOrgRep()
	nh := noteHandle{
		client:       l.ghc,
		log:          log,
		addLabels:    needAddLabels,
		removeLabels: needRemoveLabels,
		org:          org,
		repo:         repo,
		h: heh,
	}
	return neh.handleComment()

File: prow/gitee-plugins/label/label.go [hunk 8]
```diff
@@ -122,30 +122,100 @@ func (l *label) handleNoteEvent(e *sdk.NoteEvent, log *logrus.Entry) error {
...
-		neh = &prNoteHandle{noteHandle: nh, number: number}
-	} else if ne.IsIssue() {
-		number := gitee.NewIssueNoteEvent(e).GetIssueNumber()
-		neh = &issueNoteHandle{noteHandle: nh, number: number}
+func (l *label) handleComment(e gitee.NoteEventWrapper, addMatches, rmMatches [][]string, log *logrus.Entry) error {
+	org, repo := e.GetOrgRep()
+	repoLabelsMap, err := l.getRepoLabels(org, repo)
+	if err != nil {
+		return err
+	}
```
Comment (zengchen1024) [inline prow/gitee-plugins/label/label.go:133]: move 130~133 to below line 157

**Pull request merged**